### ISSUE TITLE
commit message hilight

### ISSUE THREAD

**Issue by chris-findlay** [gitextensions/gitextensions#5949]
**Question**

**What is the current behavior?**
The commit message gets a yellow hilight after 50 chars or 1 linebreak.  Where do I get to configure the colours/column limit?

I tried searching in the User Manual, but it's borken at the moment...

**Environment you encounter the issue:**
- Git Extensions 3.00.00.4433
- Build fca7cf228b481ee8c1b779cf7b882ccdfbdcd1bc
- Git 2.20.1.windows.1
- Microsoft Windows NT 10.0.17134.0
- .NET Framework 4.7.3260.0
- DPI 96dpi (no scaling)

**Comment by vbjay** [comment]
Start in this code.
https://github.com/vbjay/gitextensions/blob/9a490bfb3d7d85cf316a9daf256eb2c2703983b1/GitUI/CommandsDialogs/FormCommit.cs#L2623

**Comment by RussKie** [comment]
I believe the highlighting is done here:
https://github.com/gitextensions/gitextensions/blob/f575439c38ed8ab433bbf20f9f87659eff138db6/GitUI/SpellChecker/EditNetSpell.cs#L314-L343

The actual setting is toggled via a context menu on the `FormCommit`:
![image](https://user-images.githubusercontent.com/4403806/50203178-bd833500-03b4-11e9-9a21-285dfce572ef.png)

And it appears the limit is hardcoded and can't be configured.

**Comment by chris-findlay** [comment]
Thanx.  Now I understand the pattern it makes more sense, but in combination with the option that ensures the 2nd line is blank (i.e. a 2 line message entered becomes 3 by the insertion of the blank line) the hilight for the 2nd line is wrong.
Either way, being able to configure this (including what it looks like) would be great.

**Comment by RussKie** [comment]
I agree. The original code probably dates a number of years back, and may
do with a refactor/cleanup.

In the grand scheme of things this request is unlikely to have a high
priority stuck behind of a number of other burning issues.

A PR is certainly welcome, if you wish to have a go at it.

**Comment by drewnoakes** [comment]
> in combination with the option that ensures the 2nd line is blank (i.e. a 2 line message entered becomes 3 by the insertion of the blank line) the hilight for the 2nd line is wrong.

I am not following this, sorry. Can you paste a screenshot showing what you mean? Are you talking about a line so long it wraps in the editor? That shouldn't make a difference.

Regarding making this configurable, note that the (currently constant) value is also used in the editor form (i.e. the one that appears if you set GE as your `core.editor` and do `git commit` from the command line).

**Comment by chris-findlay** [comment]
I remember an option that has the effect of taking a 2-line commit message and inserting a blank like as line 2 so that there's always a blank line after the 'subject'.  This code doesn't look at that option, so hilights all of line 2 even if it will become line 3.

**Comment by Bejoscha** [comment]
Just chiming in here, as I was about to post an issue myself. The highlighting of long commit comments is rather irritating and confused me a lot when I first saw it after updating GITExtensions to 3 today.  After all, the "box" to type is large and there seems no real reason to keep message  super-short. 

I would very much prefer this highlighting to be an option (with a default of "off").

**Comment by vbjay** [comment]
It helps you follow 50/72 rule. So no. I don't think we will default to
off. I can see an option to turn it off. But not default.

https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html

**Comment by RussKie** [comment]
@gitextensions/git-extensions-source that's where we'd probably want to show the vertical rules by default and set at 72 char position.

Short-mid term, the setting to show/hide the ruler should be added to the context menu next to the "Mark ill formed lines":

```
...
-----
Mark ill formed lines
Hide vertical rules
Configure vertical ruler...
```

Long term we'd want to consider how the UI/UX can be improved to make settings more accessible/discoverable/configurable.

**Comment by Bejoscha** [comment]
> 
> It helps you follow 50/72 rule. So no. I don't think we will default to off. I can see an option to turn it off. But not default. https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html

Never heard of the 50/72 rule before and had to read up on it now. Everyday something new to learn - I like it. The highlighting makes a lot more sense to me now, but optional would still be nice.

![s](https://user-images.githubusercontent.com/6987920/50596105-5c269000-0ea3-11e9-96af-80fd0ddbceda.jpg)

**Comment by vbjay** [comment]
![image](https://user-images.githubusercontent.com/3629489/50596476-6da37400-0e72-11e9-9415-f8b5ef142887.png)

**Comment by Bejoscha** [comment]
@vbjay Ah! Well hidden feature then :c)
I wonder if the menu entry "Mark ill formed lines" should be called differently, though.

Also, isn't it a bit over-kill to have the "Commit templates" section _and_ this auto-enforcing of the 50/72 rule if there is no template selected? Wouldn't it be more logical to have a 50/72 template added by default ?

**Comment by vbjay** [comment]
Correction:

![image](https://user-images.githubusercontent.com/3629489/50609894-123bab00-0e9f-11e9-87f1-39b0a89bdaf1.png)
This part does control the separate pieces. The issue is that they are not template specific.  They are global control of the commit validation 

https://github.com/vbjay/gitextensions/blob/9a490bfb3d7d85cf316a9daf256eb2c2703983b1/GitUI/CommandsDialogs/CommitDialog/FormCommitTemplateSettings.cs#L59-L69

**Comment by vbjay** [comment]
With these settings
![image](https://user-images.githubusercontent.com/3629489/50610223-2b912700-0ea0-11e9-97ff-a92a970a65db.png)

- It is impossible to have text in 2nd line  It automatically skips the 2nd line
- The text is marked in red if it is invalid.

With these 
![image](https://user-images.githubusercontent.com/3629489/50610341-a65a4200-0ea0-11e9-819f-d6c413629eb9.png)

- It highlights 50 72 rules 
![image](https://user-images.githubusercontent.com/3629489/50610390-c8ec5b00-0ea0-11e9-8fc8-9c740ef597d7.png)

**Comment by Bejoscha** [comment]
Sure, but isn't a "form" that requires a first line text of <50 and then subsequent lines (after an empty one) to be <72 characters long in itself also a (minimalistic) template?

-----Original Message-----

**Comment by vbjay** [comment]
Look.  50/72 comes from git history.  Think  What is the standard console width?  What is a common print out width?  80 chars.

It is a defacto standard.  It guides developers to be short and sweet on the subject (think git log --oneline).  Git rebase doesn't like a value in 2nd line with auto squash and such.  It also helps when reading so the other defacto statndard is no text in 2nd line.  72 in subsequent lines = a margin for notes on the side of the printout.  Please quit complaining about a system that is designed to guide people to make good commit messages.

**Comment by vbjay** [comment]
Actually 72 comes from this.  Who knew?!

Git is smart enough to distinguish the first line of your commit message as your summary. In fact, if you try git shortlog, instead of git log, you will see a long list of commit messages, consisting of the id of the commit, and the summary only. The reason for wrapping your description lines at the 72nd mark is that git log adds a padding of 4 blank spaces when displaying the commit message. To center the message perfectly, and make it read well on an 80-column terminal, you’d want to leave space for 4 more blank spaces at right side. That’s where 72 comes from. https://medium.com/@preslavrachev/what-s-with-the-50-72-rule-8a906f61f09c

**Comment by vbjay** [comment]
You are confusing template 
![image](https://user-images.githubusercontent.com/3629489/50612107-d9073900-0ea6-11e9-82b6-1086b5a10aad.png)
 with commit validation rules.  I agree that the dialog could be designed better to define this separation.  

Here is how the template works

1. Define template 
![image](https://user-images.githubusercontent.com/3629489/50612188-1cfa3e00-0ea7-11e9-9f19-fe86842dca2d.png)
2. Choose template when committing.  
![image](https://user-images.githubusercontent.com/3629489/50612232-42874780-0ea7-11e9-9c9c-6efdc0b24c1a.png)
3. Save time typing 
![image](https://user-images.githubusercontent.com/3629489/50612252-5b8ff880-0ea7-11e9-88d2-a108bf93fecb.png)

**Comment by vbjay** [comment]
The validation rules are GLOBAL to the application and have NOTHING to do with specific templates.  Look at the settings in teh settings file.

```xml
<item>
    <key>
      <string>CommitTemplates</string>
    </key>
    <value>
      <string>[{"Name":"Test","Text":"xxx"},{"Name":"","Text":""},{"Name":"","Text":""},{"Name":"","Text":""},{"Name":"","Text":""}]</string>
    </value>
  </item>
  <item>
    <key>
      <string>CommitValidationAutoWrap</string>
    </key>
    <value>
      <string>true</string>
    </value>
  </item>
  <item>
    <key>
      <string>CommitValidationIndentAfterFirstLine</string>
    </key>
    <value>
      <string>false</string>
    </value>
  </item>
  <item>
    <key>
      <string>CommitValidationMaxCntCharsFirstLine</string>
    </key>
    <value>
      <string>10</string>
    </value>
  </item>
  <item>
    <key>
      <string>CommitValidationMaxCntCharsPerLine</string>
    </key>
    <value>
      <string>0</string>
    </value>
  </item>
  <item>
    <key>
      <string>CommitValidationSecondLineMustBeEmpty</string>
    </key>
    <value>
      <string>true</string>
    </value>
  </item>
  <item>
```

**Comment by vbjay** [comment]
Notice groupbox and where the border of the groupbox is.  
![image](https://user-images.githubusercontent.com/3629489/50612620-8f1f5280-0ea8-11e9-8674-3c51fa627894.png)

**Comment by vbjay** [comment]
Notice the properties of a TemplateItem.  https://github.com/gitextensions/gitextensions/blob/b210331fb67b2362cc1c666ad49911d4666d85e0/GitCommands/CommitTemplateItem.cs#L13-L14

**Comment by vbjay** [comment]
When you have validation rules set and try to commit

![image](https://user-images.githubusercontent.com/3629489/50614535-21c2f000-0eaf-11e9-8433-46e729ac44df.png)

![image](https://user-images.githubusercontent.com/3629489/50614595-559e1580-0eaf-11e9-9a22-796ac9df5063.png)

**Comment by vbjay** [comment]
Fyi. The yellow highlight doesn't stop you from commiting. The red is a validation failure that will ado you are you sure?

**Comment by Bejoscha** [comment]
@vbjay Thanks for the detailed description. It makes a lot more sense to me now . I don't disagree with it or complain! I can (now) see that the 50/72 rule has its valid reasons and that nudging users to use it is a good thing. However, the problem is, that the UI design fails a bit at what a UI is supposed to do: Intuitively guide a user instead of creating confusion. 

I guess anybody confused _can_ now read this thread here and become enlightened, but many will not and will just remain confused. Keep in mind that many users install GITExtensions (or similar) because they are _not_ all that experienced in GIT and want to _simplify_ their experience. This was certainly true for me when I did it.

**If it is all about having the _functionality_ in GITExtensions, then I think we're fine as we are.** 
If it is, however, also about user-convenience/simplification then there might be some room for improvements. 

Some simple things that jump to mind:

- Don't start out with 5 empty templates on install. Have at least one "pre-made" template so that the user can just "experience" what it is for. Very similar to the example you've actually posted above.

- The red text is visually less aggressive than the yellow highlight. This is a bit of a missmatch to "red is a serious validation failure" vs. "yellow is just an informative highlight". 

- If the yellow highlight always enforces 50/72, then the according menu entry should possibly be called like that, i.e. _"Mark ill fomatted lines"_ could be _"Highlight 50/72 format rule"_. This would give users without the knowledge a starting point.

- Minor (separate) bug: When the settings are changed, the highlights are not recomputed on exiting the dialog.

- If the validation rules are _stricter_ than the 50/72, the visuals aren't that great. Does such a double-highlighting fulfill any practical purpose?
![image](https://user-images.githubusercontent.com/6987920/50647458-e12fa900-0f78-11e9-8706-818fb209f8fd.png)

And, returning to my intial thoughts: Is there any real benefit in having both? I mean, doesn't setting the global verification rules to 50/72 do the trick already? And if that setting has been changed by the user - why would he then want/need a highlight for that rule?

**Comment by vbjay** [comment]
Yes.  It has room for improvement.  On the priority scale it kind of fell to a low level.  Pull requests are welcome.  The only requirement I would suggest to have is not removal of the feature but improvement on its use. 

> Minor (separate) bug: When the settings are changed, the highlights are not recomputed on exiting the dialog.

File as a separate issue.  I noticed the same thing.  

> If the validation rules are stricter than the 50/72, the visuals aren't that great. Does such a double-highlighting fulfill any practical purpose?  

If they set line limits (non 0 values) then 50/72 highlight should not be shown.  File as separate issue.  

> The red text is visually less aggressive than the yellow highlight.


The idea is red= error where yellow is a warning.  

> Don't start out with 5 empty templates on install. Have at least one "pre-made" template so that the user can just "experience" what it is for.  Pull requests are welcome.  File as an issue and make it happen.

**Comment by vbjay** [comment]
Boy.  That yellow and red look awesome together.  :D  NOT!

**Comment by vbjay** [comment]
2nd thought.  

> If the validation rules are stricter than the 50/72, the visuals aren't that great. Does such a double-highlighting fulfill any practical purpose?

- Instead of red text, a red highlight like the yellow
- Use an error provider if user set validation rules(non 50/72 but would be red text)  Give them info on why it is wrong.  
-Maybe suggest an autofix feature?  Autowrap your lines if too long.
- That brings up the idea of markdown esque editing.  User selects lines and hits a bullet point line and it creates 
```
- blaaa
- blaaa
...
```
lines.

**Comment by grafanauser** [comment]
Found this issue by googleing for this strange behavior. This yellow highlight is really confusing. It looks like a Bug. The user needs feedback from the GUI to understand why this is happening.

**Comment by pmiossec** [comment]
>The user needs feedback from the GUI to understand why this is happening.

What do you propose?

I don't see a way to improve it without making the GUI cumbersome. But perhaps you have a good idea.

I'm not very fond of the yellow, personally and I think that we could change it for a very light blue similar to the selected line blue so that user better understand that it is linked....

**Comment by pmiossec** [comment]
>I'm not very fond of the yellow, personally and I think that we could change it for a very light blue similar to the selected line blue so that user better understand that it is linked....

I mean, by default, because that's configurable...

**Comment by grafanauser** [comment]
The color doesnt matter. It really looks like a graphical glitch, when you dont know the purpose of it. Is it possible to add a tooltip to the highlighted part when the mouse is above it? Or a tooltip to the whole commit message textbox "Yellow highlighted text indicates a deviation from the 50/70 rule for commit messages".  Or open a popup once, when this happens for the very first time. Though popups can be quite annoying.

**Comment by pmiossec** [comment]
Sorry, I misunderstood the context and I thought this issue was talking about something else...

I agree that it could be improved. A tooltip could indeed be a good idea.
We could also display a link toward a post on how to write a good commit message (to improve everyone knowledge!) and display it only when the yellow mark is shown. But I suppose such idea won't please everyone ;)

**Comment by NamelessVoice** [comment]
Could definitely do with some improvement, it really does look like a bug the first time you see it.

A little info icon with a tooltip telling you why it's being highlighted and indicating how to configure/turn it off if you don't care about the rule would be nice.

**Comment by lukevp** [comment]
How about the default settings becoming 50 and 72 so that it is not functionally a regression for current users, but a setting of 0 would disable the highlighting?

**Comment by gerhardol** [comment]
The setting should not be changed now, it would only affect those that upgrade from 2.5x at this point, which I would assume is a minority (no telemetry to back that).

If something is changed, it should be the tooltip or so.

**Comment by lukevp** [comment]
@gerhardol Thanks for the input.  I've looked into this and the yellow highlighting is occurring in the spell check function.  I feel that the highlighter should respect the same settings rather than having hardcoded 50 and 72 built into it, which lets the user configure this to their own preference.  
I've opened a PR that makes this change, let me know what you think.  PR #6774

**Comment by gusarov** [comment]
Just to bump up this topic: Our company policy is 72/72, so for both first and next lines. Can we have a setting to configure highlighting rules? I think it is beneficial to have one, rules are changing over time, screens are getting bigger, tasks and messages are getting more complex and complicated e.t.c.
